app/models/sample_manifest/uploader.rb

```diff
@@ -28,19 +28,22 @@ def initialize(file, configuration, user, override)
   end
 
   def run!
-    ActiveRecord::Base.transaction do
-      return false unless valid? # rubocop:todo Rails/TransactionExitStatement
-
-      return true if process_upload_and_callbacks # rubocop:todo Rails/TransactionExitStatement
-
-      # One of our post processing checks failed, something went wrong, so we
-      # roll everything back
-      raise ActiveRecord::Rollback
-    end
-
-    extract_errors
-    upload.fail
-    false
+    result =
```

Comment:
what is the reason for the change here? Not entirely sure what is going on.

Reply:
This is actually something I find difficult to reason. Take the previous version of the code (that was there in Rails 6.0):

```ruby
 def run!
    ActiveRecord::Base.transaction do
      return false unless valid? # rubocop:todo Rails/TransactionExitStatement

      return true if process_upload_and_callbacks # rubocop:todo Rails/TransactionExitStatement

      # One of our post processing checks failed, something went wrong, so we
      # roll everything back
      raise ActiveRecord::Rollback
    end

    extract_errors
    upload.fail
    false
  end
```

It appears the line `return true if process_upload_and_callbacks` throws out an exception in Rails 6.1 (not a very helpful exception, I'm afraid):

```
 Failure/Error:
       ActiveRecord::Base.transaction do
         return false unless valid? # rubocop:todo Rails/TransactionExitStatement
   
         return true if process_upload_and_callbacks # rubocop:todo Rails/TransactionExitStatement
   
         # One of our post processing checks failed, something went wrong, so we
         # roll everything back
         raise ActiveRecord::Rollback
       end
 
 ActiveSupport::DeprecationException:
   DEPRECATION WARNING: Using `return`, `break` or `throw` to exit a transaction block is
   deprecated without replacement. If the `throw` came from
   `Timeout.timeout(duration)`, pass an exception class as a second
   argument so it doesn't use `throw` to abort its block. This results
   in the transaction being committed, but in the next release of Rails
   it will rollback.
    (called from run! at /Users/dp27/Documents/Projects/sequencescape/app/models/sample_manifest/uploader.rb:31)
 # ./app/models/sample_manifest/uploader.rb:31:in `run!'
 # ./app/controllers/sample_manifest_upload_with_tag_sequences_controller.rb:13:in `create'
 # ./lib/capybara_timeout_patches.rb:10:in `synchronize'
 # ./spec/features/sample_manifests/uploader_for_manifests_with_tag_sequences_spec.rb:208:in `block (5 levels) in <top (required)>'
 # ./spec/spec_helper.rb:87:in `block (3 levels) in <top (required)>'
 # ./spec/spec_helper.rb:87:in `block (2 levels) in <top (required)>'
```

I've noticed that this issue occur only when returning true within the transaction scope (returning within a transaction is not recommended and is said to be deprecated in future Rails releases). My solution was to extract out the result from the transaction, and handle it as I've done it in the changes.

@yoldas do you have any suggestions on this?

Reply:
The following is a good quick summary and lists future support issues: https://www.loyalty.dev/posts/returning-from-transactions-in-rails

The existing code is trying to return from the method inside a transaction block. That needs changing. 

As discussed, valid? check can be put outside of the transaction block as an initial guard statement in the method. 

The transaction block can generate a value and it can be assigned to a variable to use in guard statement to skip the calls at the end.

Note that raising a Rollback exception inside the transaction block does not bubble up, it is used as a sign to rollback the current transaction by the block.

There must not be any return statement inside the transaction block; independent from return value. See how using returns inside the transaction block behaves in different Rails versions at the link above.